### PR TITLE
fix: Fix typing for client_options as dict

### LINKED ISSUE #1380: Dictionary support not reflected in type annotations
## Issue

- This is a follow-up to [#986](https://github.com/googleapis/gapic-generator-python/issues/986) (closed). Retesting the recent `google-cloud-documentai==1.5.0` version, the issue is still present.
- This code is valid:
  ```
  import google.cloud.documentai_v1 as documentai

  client_options = dict(api_endpoint="eu-documentai.googleapis.com")
  client = documentai.DocumentProcessorServiceClient(client_options=client_options)
  ```
- But it triggers the following type checker warning:
  ```
  Argument of type "dict[str, str]" cannot be assigned to parameter "client_options" of type "ClientOptions | None" in function "__init__"
  Type "dict[str, str]" cannot be assigned to type "ClientOptions | None"
  "dict[str, str]" is incompatible with "ClientOptions"
  ```
- If you check the [source](https://github.com/googleapis/python-documentai/blob/1b4eff4e0f6ca34bd12fb5cd01343ca74b4f9633/google/cloud/documentai_v1/services/document_processor_service/client.py#L366), `client_options` can be a dict, yet its declaration says otherwise:
  ```python
  def __init__(
      ...
      client_options: Optional[client_options_lib.ClientOptions] = None,
      ...
  ) -> None:
      ...
      if isinstance(client_options, dict):
          client_options = client_options_lib.from_dict(client_options)
  ```

## Expected

  ```python
  def __init__(
      ...
      client_options: Union[client_options_lib.ClientOptions, dict, None] = None,
      ...
  ) -> None:
      ...
  ```

### PR DESCRIPTION
**Pull request by PicardParis**
Fixes #1380

### CODE REVIEW

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 1]
```diff
@@ -276,7 +276,7 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
     def __init__(self, *,
             credentials: Optional[ga_credentials.Credentials] = None,
             transport: Union[str, {{ service.name }}Transport, None] = None,
-            client_options: Optional[client_options_lib.ClientOptions] = None,
+            client_options: Union[client_options_lib.ClientOptions, dict, None] = None,
```
Comment (dandhlee) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:279]: Changing from `Optional` to `Union` changes its meaning - it's the same thing technically, but I'd prefer if we could stick to using `Optional[]` than `Union[..., None]`.

```suggestion
            client_options: Optional[Union[client_options_lib.ClientOptions, dict]] = None,
```

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 2]
```diff
@@ -295,7 +295,7 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
                 beta state (preview). We welcome your feedback via an
                 issue in this library's source repository.
                 {% endif %}
-            client_options (google.api_core.client_options.ClientOptions): Custom options for the
+            client_options (Union[google.api_core.client_options.ClientOptions, dict, None]): Custom options for the
```
Comment (dandhlee) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:298]: Ditto:

```suggestion
            client_options (Optional[Union[google.api_core.client_options.ClientOptions, dict]]): Custom options for the
```

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 (hunk 1 above)
Comment (PicardParis) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:279]: I've read that it was the preferred way as it increases the readability (which I agree with). I'll search where tomorrow if that matters...

Comment (PicardParis) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:279]: This was also accepted by @atulep in [#1380](https://github.com/googleapis/gapic-generator-python/issues/1380#issuecomment-1215886077). Let me know...

Comment (atulep) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:279]: I am okay with this. Seems like we already have done something similar for `transport` argument: `transport: Union[str, {{ service.name }}Transport, None] = None`.

**Comment by PicardParis** [comment]
- This should be consistent now.
- @dandhlee / Out of curiosity, what will be our convention when 3.10 is the min supported version:
  - `client_options: Optional[client_options_lib.ClientOptions | dict] = None,`
  - `client_options: client_options_lib.ClientOptions | dict | None = None,`

**Review by holtskinner — APPROVED** [review]

**Comment by dandhlee** [comment]
Unless there's a huge performance implication for using 
```
client_options: Optional[client_options_lib.ClientOptions | dict] = None,
```
I think that's still more straightforward and direct than 
```
client_options: client_options_lib.ClientOptions | dict | None = None,
```

that said, if the general consensus of the Python community changes then we'd adapt. For now, we might utilize the `|` type hint operations added but prefer the former to note that it's `Optional`.

**Comment by PicardParis** [comment]
Hi! Can it be merged? Should anything else be done? (this is my first PR to this repo)

**Comment by atulep** [comment]
Hi @PicardParis 

Some CI tests are failing. Try upgrading the branch with main - I think it should solve some of the issues.

**Comment by PicardParis** [comment]
Hi @atulep,
I clicked _"Update branch"_. The state is now _"Merging can be performed automatically with 1 approving review"_.

**Comment by PicardParis** [comment]
Hi @vam-google @parthea @atulep,
The branch is up-to-date.
Can you rerun the checks? (I'm a first-time contributor and can only push new commits)

**Pull request closed without merging**